File: Source/Back-end/src/main/java/org/elis/manutenzione/service/definition/LuogoService.java
package org.elis.manutenzione.service.definition;

import org.elis.manutenzione.dto.request.AggiungiLuogoRequest;
import org.elis.manutenzione.dto.request.EliminaLuogoRequest;
import org.elis.manutenzione.dto.request.ModificaStanzaRequest;
import org.elis.manutenzione.model.entity.Luogo;
import org.elis.manutenzione.model.entity.Residente;

import java.util.List;

/**
 * Interfaccia del servizio per la gestione dei luoghi.
 * Questa interfaccia definisce i metodi per le operazioni relative ai luoghi,
 * come l'aggiunta, la modifica, l'eliminazione e il recupero dei luoghi.
 */
public interface LuogoService {

    /**
     * Aggiunge un nuovo luogo nel sistema.
     *
     * @param request DTO contenente i dati del luogo da aggiungere.
     * @return L'oggetto Luogo appena creato e salvato.
     */
    public Luogo aggiungiLuogo(AggiungiLuogoRequest request);



    /**
     * Recupera la lista di tutti i luoghi presenti nel sistema.
     *
     * @return La lista di tutti gli oggetti Luogo.
     */
    public List<Luogo> getLuoghi();


    /**
     * Elimina un luogo specifico dal sistema.
     *
     * @param request DTO contenente l'id del luogo da eliminare.
     * @return true se l'operazione di eliminazione è completata con successo, false altrimenti.
     */
    public boolean eliminaLuogo(EliminaLuogoRequest request);


    /**
     * Recupera la lista di tutti i luoghi comuni.
     *
     * @return La lista di tutti gli oggetti Luogo di tipo LuogoComune.
     */
    public List<Luogo> getLuoghiComuni();


    /**
     * Recupera la lista di tutte le stanze.
     *
     * @return La lista di tutti gli oggetti Luogo di tipo Stanza.
     */
    public List<Luogo> getStanze();


    /**
     * Recupera la lista di tutti i bagni.
     *
     * @return La lista di tutti gli oggetti Luogo di tipo Bagno.
     */
    public List<Luogo> getBagni();


    /**
     * Modifica la capienza di una stanza specifica.
     *
     * @param request DTO contenente l'id della stanza e la nuova capienza.
     * @return L'oggetto Luogo (stanza) modificato.
     */
    public Luogo modificaStanza(ModificaStanzaRequest request);


    /**
     * Recupera la lista dei luoghi associati ad un residente specifico.
     *
     * @param r Il residente di cui si vogliono recuperare i luoghi.
     * @return La lista di tutti gli oggetti Luogo associati al residente.
     */
    public Luogo getStanzaByResidente(Residente r);
}

File: Source/Front-end/front-end/src/app/services/luogo/luogo.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from "@angular/common/http";
import {environment} from "../../../environments/environment";
import {Observable} from "rxjs";
import {AggiungiLuogoRequest} from "../../model/request/aggiungi-luogo-request";
import {AggiungiLuogoResponse} from "../../model/response/aggiungi-luogo-response";
import {GetLuoghiResponse} from "../../model/response/get-luoghi-response";
import {GetStanzeResponse} from "../../model/response/get-stanze-response";
import {GetLuoghiComuniResponse} from "../../model/response/get-luoghi-comuni-response";
import {GetBagniResponse} from "../../model/response/get-bagni-response";
import {EliminaLuogoRequest} from "../../model/request/elimina-luogo-request";
import {ModificaStanzaRequest} from "../../model/request/modifica-stanza-request";
import {Luogo} from "../../model/models/luogo";
import {GetStanzaDto} from "../../model/models/get-stanza-dto";

@Injectable({
  providedIn: 'root'
})
export class LuogoService {

  constructor(private http: HttpClient) {
  }

  /**
   * Invia la richiesta di aggiungi luogo al BE
   */
  aggiungiLuogo(aggiungiLuogoRequest: AggiungiLuogoRequest): Observable<AggiungiLuogoResponse> {
    return this.http.post<AggiungiLuogoResponse>(environment.baseUrl + environment.Admin.baseUrl + environment.Luogo.aggiungiLuogo, aggiungiLuogoRequest);
  }

  /**
   * Invia la richiesta di ritrovamento luoghi al BE
   */
  getLuoghi(): Observable<GetLuoghiResponse[]> {
    return this.http.get<GetLuoghiResponse[]>(environment.baseUrl + environment.All.baseUrl + environment.Luogo.getLuoghi);
  }

  /**
   * Invia la richiesta di ritrovamento stanze al BE
   */
  getStanze(): Observable<GetStanzeResponse[]> {
    return this.http.get<GetStanzeResponse[]>(environment.baseUrl + environment.Admin.baseUrl + environment.Luogo.getStanze);
  }

  /**
   * Invia la richiesta di ritrovamento luoghi comuni al BE
   */
  getLuoghiComuni(): Observable<GetLuoghiComuniResponse[]> {
    return this.http.get<GetLuoghiComuniResponse[]>(environment.baseUrl + environment.Admin.baseUrl + environment.Luogo.getLuoghiComuni);
  }

  /**
   * Invia la richiesta di ritrovamento bagni al BE
   */
  getBagni(): Observable<GetBagniResponse[]> {
    return this.http.get<GetBagniResponse[]>(environment.baseUrl + environment.Admin.baseUrl + environment.Luogo.getBagni);
  }

  /**
   * Invia la richiesta di elimina luogo al BE
   */
  eliminaLuogo(eliminaLuogoRequest: EliminaLuogoRequest): Observable<boolean> {
    return this.http.delete<boolean>(environment.baseUrl + environment.Admin.baseUrl + environment.Luogo.eliminaLuogo, {
      body: eliminaLuogoRequest
    });
  }

  /**
   * Invia la richiesta di modifica stanza al BE
   */
  modificaStanza(modificStanzaRequest: ModificaStanzaRequest): Observable<Luogo[]> {
    return this.http.patch<Luogo[]>(environment.baseUrl + environment.Admin.baseUrl + environment.Luogo.modificaStanza, modificStanzaRequest);
  }

  /**
   * Invia la richiesta di ritrovamento stanza del residente al BE
   */
  getStanza(){
    return this.http.get<GetStanzaDto>(environment.baseUrl + environment.Residente.baseUrl + environment.Luogo.getStanza);
  }
}
